packages/realm/src/List.ts

```diff
@@ -343,3 +312,171 @@ export class List<T = unknown> extends OrderedCollection<T> implements Partially
+  };
+}
+
+function getMixed<T>(realm: Realm, typeHelpers: TypeHelpers<T>, list: binding.List, index: number): T {
+  const value = list.getAny(index);
+  switch (value) {
+    case binding.ListSentinel: {
+      const accessor = createListAccessor<T>({ realm, typeHelpers, itemType: binding.PropertyType.Mixed });
+      return new List<T>(realm, list.getList(index), accessor) as T;
+    }
+    case binding.DictionarySentinel: {
+      const accessor = createDictionaryAccessor<T>({ realm, typeHelpers, itemType: binding.PropertyType.Mixed });
+      return new Dictionary<T>(realm, list.getDictionary(index), accessor) as T;
+    }
+    default:
+      return typeHelpers.fromBinding(value);
+  }
+}
+
+function setKnownType<T>(
+  realm: Realm,
+  toBinding: TypeHelpers<T>["toBinding"],
+  isEmbedded: boolean,
+  list: binding.List,
+  index: number,
+  value: T,
+): void {
+  assert.inTransaction(realm);
+  list.setAny(index, toBinding(value, isEmbedded ? { createObj: () => [list.setEmbedded(index), true] } : undefined));
+}
+
+function setMixed<T>(
+  realm: Realm,
+  toBinding: TypeHelpers<T>["toBinding"],
+  list: binding.List,
+  index: number,
+  value: T,
+): void {
+  assert.inTransaction(realm);
+
+  if (isJsOrRealmList(value)) {
+    list.setCollection(index, binding.CollectionType.List);
+    insertIntoListOfMixed(value, list.getList(index), toBinding);
+  } else if (isJsOrRealmDictionary(value)) {
+    list.setCollection(index, binding.CollectionType.Dictionary);
+    insertIntoDictionaryOfMixed(value, list.getDictionary(index), toBinding);
+  } else {
+    list.setAny(index, toBinding(value));
+  }
+}
+
+function insertKnownType<T>(
+  realm: Realm,
+  toBinding: TypeHelpers<T>["toBinding"],
+  isEmbedded: boolean,
+  list: binding.List,
+  index: number,
+  value: T,
+): void {
+  assert.inTransaction(realm);
+
+  if (isEmbedded) {
+    // Simply transforming to binding will insert the embedded object
+    toBinding(value, { createObj: () => [list.insertEmbedded(index), true] });
+  } else {
+    list.insertAny(index, toBinding(value));
+  }
+}
+
+function insertMixed<T>(
+  realm: Realm,
+  toBinding: TypeHelpers<T>["toBinding"],
+  list: binding.List,
+  index: number,
+  value: T,
+): void {
+  assert.inTransaction(realm);
+
+  if (isJsOrRealmList(value)) {
+    list.insertCollection(index, binding.CollectionType.List);
+    insertIntoListOfMixed(value, list.getList(index), toBinding);
+  } else if (isJsOrRealmDictionary(value)) {
+    list.insertCollection(index, binding.CollectionType.Dictionary);
+    insertIntoDictionaryOfMixed(value, list.getDictionary(index), toBinding);
+  } else {
+    list.insertAny(index, toBinding(value));
+  }
+}
+
+/** @internal */
+export function insertIntoListOfMixed(
+  list: List | unknown[],
+  internal: binding.List,
+  toBinding: TypeHelpers["toBinding"],
+) {
+  // TODO: Solve the "removeAll()" case for self-assignment.
+  internal.removeAll();
+
+  let index = 0;
+  for (const item of list) {
```

Comment:
Is `for (const [index, item] of list.entries)` possible here?

Reply:
Yes that'd be possible with `list.entries()` 🙂 And now that I think about it, it may even be safer to do it that way because if the `list` is a `Realm.List`, then `list.entries()` would take the snapshot.